logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/e2ei/EnrolE2EIUseCase.kt

```diff
@@ -0,0 +1,138 @@
+            return E2EIEnrolmentResult.Failed(E2EIEnrolmentResult.E2EIStep.AcmeNewOrder, it).toEitherLeft()
+        }, { it })
+
+        prevNonce = newOrderResponse.second
+
+        val authzResponse = e2EIRepository.createAuthz(prevNonce, newOrderResponse.first.authorizations[0]).fold({
+            return E2EIEnrolmentResult.Failed(E2EIEnrolmentResult.E2EIStep.AcmeNewAuthz, it).toEitherLeft()
+        }, { it })
+
+        prevNonce = authzResponse.second
+
+        val wireNonce = e2EIRepository.getWireNonce().fold({
+            return E2EIEnrolmentResult.Failed(E2EIEnrolmentResult.E2EIStep.WireNonce, it).toEitherLeft()
+        }, { it })
+
+        val dpopToken = e2EIRepository.getDPoPToken(wireNonce).fold({
+            return E2EIEnrolmentResult.Failed(E2EIEnrolmentResult.E2EIStep.DPoPToken, it).toEitherLeft()
+        }, { it })
+
+        val wireAccessToken = e2EIRepository.getWireAccessToken(dpopToken).fold({
+            return E2EIEnrolmentResult.Failed(E2EIEnrolmentResult.E2EIStep.WireAccessToken, it).toEitherLeft()
+        }, { it })
+
+        val dpopChallengeResponse = e2EIRepository.validateDPoPChallenge(
+            wireAccessToken.token, prevNonce, authzResponse.first.wireDpopChallenge!!
+        ).fold({
+            return E2EIEnrolmentResult.Failed(E2EIEnrolmentResult.E2EIStep.DPoPChallenge, it).toEitherLeft()
+        }, { it })
+
+        prevNonce = dpopChallengeResponse.nonce
+
+        val oidcChallengeResponse = e2EIRepository.validateOIDCChallenge(
+            idToken, prevNonce, authzResponse.first.wireOidcChallenge!!
+        ).fold({
+            return E2EIEnrolmentResult.Failed(E2EIEnrolmentResult.E2EIStep.OIDCChallenge, it).toEitherLeft()
+        }, { it })
+
+        prevNonce = oidcChallengeResponse.nonce
+
+        val orderResponse = e2EIRepository.checkOrderRequest(newOrderResponse.third, prevNonce).fold({
+            return E2EIEnrolmentResult.Failed(E2EIEnrolmentResult.E2EIStep.CheckOrderRequest, it).toEitherLeft()
+        }, { it })
+
+        prevNonce = orderResponse.first.nonce
+
+        // TODO(fix): replace with orderResponse.third
+        val finalizeResponse = e2EIRepository.finalize(orderResponse.second, prevNonce).fold({
+            return E2EIEnrolmentResult.Failed(E2EIEnrolmentResult.E2EIStep.FinalizeRequest, it).toEitherLeft()
+        }, { it })
+
+        prevNonce = finalizeResponse.first.nonce
+
+        val certificateRequest = e2EIRepository.certificateRequest(finalizeResponse.second, prevNonce).fold({
+            return E2EIEnrolmentResult.Failed(E2EIEnrolmentResult.E2EIStep.Certificate, it).toEitherLeft()
+        }, { it })
+
+        // TODO(fix): init after fixing the MLS client initialization mechanism
+        // TODO(revert): e2EIRepository.initMLSClientWithCertificate(certificateRequest.response.decodeToString())
+        return Either.Right(E2EIEnrolmentResult.Success(certificateRequest.response.decodeToString()))
```

Comment:
TODOs needs to be done.

Reply:
this will be covered after refactoring the structure of MLSInitialisation.